contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,439 @@
+event to the watcher and then just closing this particular watch.
+
+ At the high level, we would like the API to be generic enough so that it will
+be useful for many different usecases, not just this particular one (another
+example may be a controller that needs to deal with just subset of namespaces).
+As a result, below we are describing requirements that the end solution has to
+meet to satisfy our needs
+- a single bulk requests has to support multiple resource types (e.g. get a
+node and all pods associated with it)
+- the wrappers for aggregating multiple objects (in case of list we can return
+a number of objects of different kinds) should be `similar` to lists in core
+API (by lists I mean e.g. `PodList` object)
+- the API has to be implemented also in aggregator so that bulk operations
+are supported also if different resource types are served by different
+apiservers
+- clients has to be able to alter their watch subscribtions incrementally (it
+may not be implemented in the initial version though, but has to be designed)
+
+
+# Detailed design
+
+ As stated in above requirements, we need to make bulk operations work across
+different resource types (e.g. watch pod P and secret S within a single watch
+call). To clarify, the initial implementation doesn't need to support it, but
+it has to be possible to extend the implementation without changing the API.
+Spanning multiple resources, resources types or conditions will be more and more
+important for large number of watches. As an example, federation will be adding
+watches for every type it federates. With that in mind, bypassing aggregation
+at the resource type level and going to aggregation over objects with different
+resource types will allow us to more aggresively optimize in the future (it
+doesn't mean you have to watch resources of different types in a single watch,
+but we would like to make it possible).
+
+ Moreover, our current REST API doesn't even offer an easy way to handle
+"multiple watches of a given type" within a single request. As a result, instead
+of inventing new top level pattern per type, we can introduce a new resource
+type that follows normal RESTful rules and solves even more generic problem
+of spanning multiple different resource types.
+
+ We will start with introducing a new dedicated API group:
+ ```
+ /apis/bulk.k8s.io/
+ ```
+ that underneath will have a completely separate implementation.
+
+ In all text below, we are assuming v1 version of the API, but it will obviously
+go through alpha and beta stages before (it will start as v1alpha1).
+
+ In this design, we will focus only on bulk get (list) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `getoperation` resource and supporting the
+following operation:
+```
+POST /apis/bulk.k8s.io/v1/getoperation <body defines filtering>
```

Comment:
If we're going to call it `subscriptions`, we could just call it `watches`.  However, I think `BulkGetOperation` is probably appropriate - it's not too generic, and because it supports multiple watches, the incremental "subscribe" style works as well, and leaves the door open.  I originally worried about stuttering, but I think given the novelty of this operation BulkGetOperation (for the endpoint) and GetOperation (for the incremental watch) is appropriate.